### PR TITLE
tiffload: use TIFFRGBAImage or TIFFReadRGBATile if tiff image type is not natively supported by libvips

### PR DESCRIPTION
**Pull request by DarthSim**
Hey there 👋

There are still some TIFFs with old-style JPEG (OJPEG) compression in the wild, and our users face problems decoding them. Dealing with such TIFFs using low-level libtiff methods is pretty problematic, but luckily libtiff provides high-level methods for decoding any image as 8-bit RGBA. As a bonus, we can use these methods to decode TIFFs that are not natively supported by libvips such as non-JPEG subsampled ones.

I used [`TIFFRGBAImage`](https://linux.die.net/man/3/tiffrgbaimagebegin) for scanline-wise read and [`TIFFReadRGBATile`](http://www.libtiff.org/man/TIFFReadRGBATile.3t.html) for tile-wise read.

Here's a TIFF with OJPEG compression: https://github.com/openpreserve/format-corpus/raw/master/tiff-examples/old-style-jpeg-compression/old-style-jpeg-compression.tif
And here's an archive with multiple TIFF samples including OJPEG stripwise (smallliz.tif), OJPEG tilewise (zackthecat.tif), and non-JPEG sampled (ycbcr-cat.tif) https://download.osgeo.org/libtiff/pics-3.8.0.tar.gz.

### CODE REVIEW

**Comment by DarthSim** [comment]
Fixed a possible memory leak

**Review by kleisauke — COMMENTED** [review]
In terms of code, this looks good to me. However, I'm uncertain about whether we should support TIFF images using 'old-style' JPEG compression, considering it was deprecated in 1995 (which is 28 years ago, coincidentally my birth year). Additionally, libtiff issues a warning regarding this:
https://gitlab.com/libtiff/libtiff/-/blob/5e7c59d6a950760c875c865846e57b91a64ea428/libtiff/tif_ojpeg.c#L720-723

@jcupitt WDYT?

File: libvips/foreign/tiff2vips.c [hunk 1]
```diff
@@ -2906,7 +3045,8 @@ rtiff_header_read(Rtiff *rtiff, RtiffHeader *header)
 	 * non-jpg images. We must set this here since it changes the result
 	 * of scanline_size.
 	 */
-	if (header->compression != COMPRESSION_JPEG &&
+	if (!header->read_as_rgba &&
+		header->compression != COMPRESSION_JPEG &&
```
[the remark this answers is not in the document]
Comment (DarthSim) [reply libvips/foreign/tiff2vips.c:3049]: I'm not sure it'll be clearer this way. We'll have to keep in mind that we decode OJPEG as RGBA and don't care about subsampling.

**Comment by DarthSim** [comment]
I was surprised too that you can meet OJPEG in the wild. For example, some Canon cameras use OJPEG TIFF as a base for their RAW format.

JFYI: Pillow and ImageMagick support OJPEG TIFFs

**Comment by jcupitt** [comment]
I'm amazed OJPEGs are still around too. But if they are in use, I suppose it's good to support them.

File: libvips/foreign/tiff2vips.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kleisauke) [reply libvips/foreign/tiff2vips.c:3049]: Ah, you're right. I assumed `rtiff_header_read` was only called once (then `header->read_as_rgba` would only be `FALSE` when  `header->compression == COMPRESSION_OJPEG`), but it appears to be called for each page as well.

**Comment by jcupitt** [comment]
Thank you @DarthSim, let's do this!

**Comment by jcupitt** [comment]
I added a changelog note and tried to revise the comments and a few small style things, hopefully I didn't break it.

**Comment by DarthSim** [comment]
I've tested the changes on my samples, everything works fine 👍

**Pull request merged**